explorer/interpreter/value.cpp

```diff
@@ -524,6 +525,22 @@ void Value::Print(llvm::raw_ostream& out) const {
   }
 }
 
+// Generates a 'human readable' description for an enum name.
+// For example, IntType -> 'int type'.
+static auto MakeKindName(const char* name) -> std::string {
+  return llvm::StringRef(
+             std::regex_replace(name, std::regex("(?!^)([A-Z])"), " $1"))
+      .lower();
```

Comment:
Can you elaborate why you aren't just returning the name directly? Doesn't this make it harder to figure out the enum value?

Reply:
> Can you elaborate why you aren't just returning the name directly? Doesn't this make it harder to figure out the enum value?

It does a little bit, but it also results in better looking error messages:
`Expected a type, but got parameterized entity name: Point` vs 
`Expected a type, but got ParameterizedEntityName: Point`

And also matches the behavior of `{Expression/Pattern/Statement/Declaration}KindName()`

Reply:
I'll look for someone else to comment -- for me, I'd rather see that it's an enum value and have something I can search on, rather than something I can't.

Reply:
> I'll look for someone else to comment -- for me, I'd rather see that it's an enum value and have something I can search on, rather than something I can't.

Can someone look into this?

Reply:
@geoffromer Could you look at this when you get back tomorrow?

Reply:
In general, I'm in favor of programmer-facing errors being written in plain English. (On the other hand, internal errors can and should explicitly mention internal components of the explorer.)

Reply:
I really don't like raw enumerator names in some contexts like the error messages, so maybe the solution is to have both?

For this PR, renamed XxKindName() -> XxKindDesc() both for the new and existing functions, to keep consistency. 
PTAL ;)

Reply:
I'd rather just have the enum name rather than trying to ensure that a regex replacement of the enum name yields a reasonable string. I think the enum name should remain human readable.

Reply:
> I'd rather just have the enum name rather than trying to ensure that a regex replacement of the enum name yields a reasonable string. I think the enum name should remain human readable.

The string-splitting does seem to work well enough though, and I like it more. I can add XxKindName() for all impacted enums in a follow-up PR.

Reply:
I think if you want to go back to the initial approach, without the `.def` because it's then just adding unnecessary complexity, and provide the more readable text that @geoffromer suggested, I'd be fine with that (since I think you and he will have agreed on it).

If you wanted to stick with the enum, I'd also be fine with that -- it's the approach I personally favor.

However, we should work towards an approach that at least some majority agrees with.

Reply:
> I think if you want to go back to the initial approach, without the `.def` because it's then just adding unnecessary complexity, and provide the more readable text that @geoffromer suggested, I'd be fine with that (since I think you and he will have agreed on it).
> 
> If you wanted to stick with the enum, I'd also be fine with that -- it's the approach I personally favor.
> 
> However, we should work towards an approach that at least some majority agrees with.

Sure I maybe could do one of these, but given the PR is merely adding a simple debug output aid and not like trying to solve some complicated language design issue, is it even worth spending so much time (18 days so far for this tiny PR) on it debating the space character? Is it ok to have a non 100% ideal change (from someone's point of view anyways) and then evaluate and fix if it's really not working out, in the launch-and-iterate spirit?

I actually am now quite reluctant to start doing other/bigger PRs because I feel they might take forever because of some similar minor/arbitrary thing. @chandlerc for thoughts.

Reply:
FWIW, I agree with the direction suggested by @geoffromer (and originally by @jonmeow):

> My recommendation would be to postpone diving down that rathole, and stick with enumerator names for now.

I really do understand that you don't like the resulting error messages. I agree they could be better for users. But I think that is better addressed with a holistic design, and a design that is discussed with folks working on the explorer ahead of time. As several folks have pointed out, using a regular expression to parse the enum name is going to quickly hit issues.

To get good user-facing error strings, we're going to need some form of proper diagnostics engine that formats errors in a human readable way. Clang and even bits of the toolchain can serve as ideas for these designs, but I think they will look substantially different and be a substantially larger change.

Geoff's point is to not try to solve that problem today (in this PR), and use the much simpler form of just printing the actual enum as it is spelled in the explorer codebase. That makes sense to me given the current state of things.

However, you also raised an important point:

> And also matches the behavior of {Expression/Pattern/Statement/Declaration}KindName()

But @geoffromer directly suggested those should change to also directly print the enum. IMO, that also makes sense.

So I think there are two clear paths forward here that maybe haven't been stated as explicitly as they might:

- a) Land this PR with the exact spelling of the enums. Send a follow-up PR to change the other `*KindName()` functions to be consistent.
- b) Land this PR with the spelling consistent with existing `*KindName()` functions. Send a follow-up PR to change all the `*KindName()` functions to directly return the enum names.

It seems like (b) is strictly more work than (a) so I would probably suggest (a). I think that was the implicit suggestion from both Jon and Geoff.

---

> Sure I maybe could do one of these, but given the PR is merely adding a simple debug output aid and not like trying to solve some complicated language design issue, is it even worth spending so much time (18 days so far for this tiny PR) on it debating the space character? Is it ok to have a non 100% ideal change (from someone's point of view anyways) and then evaluate and fix if it's really not working out, in the launch-and-iterate spirit?

No one is suggesting to have a 100% ideal change, they're just suggesting to avoid adding another instance of complexity that doesn't make sense to them as a code reviewer. I think there is a pretty clear path forward here.

> I actually am now quite reluctant to start doing other/bigger PRs because I feel they might take forever because of some similar minor/arbitrary thing. @chandlerc for thoughts.

I'm sorry you're reluctant, but I do think there was already a pretty clear path forward here.

I think I've mentioned this before, but I feel like discussing planned changes ahead of time, or discussing disagreements in real time in Discord or one of the meetings might help speed things up. Moreover, I think if you want to make more incremental / fast progress, it may help to move longer discussions around changing the direction suggested in code review to not be in the critical path of the code review and instead be separate discussions.

Anyways, my two cents here.

Reply:
> FWIW, I agree with the direction suggested by @geoffromer (and originally by @jonmeow):
> 
> > My recommendation would be to postpone diving down that rathole, and stick with enumerator names for now.
> 
> I really do understand that you don't like the resulting error messages. I agree they could be better for users. But I think that is better addressed with a holistic design, and a design that is discussed with folks working on the explorer ahead of time. As several folks have pointed out, using a regular expression to parse the enum name is going to quickly hit issues.
> 
> To get good user-facing error strings, we're going to need some form of proper diagnostics engine that formats errors in a human readable way. Clang and even bits of the toolchain can serve as ideas for these designs, but I think they will look substantially different and be a substantially larger change.
> 
> Geoff's point is to not try to solve that problem today (in this PR), and use the much simpler form of just printing the actual enum as it is spelled in the explorer codebase. That makes sense to me given the current state of things.
> 
> However, you also raised an important point:
> 
> > And also matches the behavior of {Expression/Pattern/Statement/Declaration}KindName()
> 
> But @geoffromer directly suggested those should change to also directly print the enum. IMO, that also makes sense.
> 
> So I think there are two clear paths forward here that maybe haven't been stated as explicitly as they might:
> 
> * a) Land this PR with the exact spelling of the enums. Send a follow-up PR to change the other `*KindName()` functions to be consistent.
> * b) Land this PR with the spelling consistent with existing `*KindName()` functions. Send a follow-up PR to change all the `*KindName()` functions to directly return the enum names.
> 
> It seems like (b) is strictly more work than (a) so I would probably suggest (a). I think that was the implicit suggestion from both Jon and Geoff.
> 
> > Sure I maybe could do one of these, but given the PR is merely adding a simple debug output aid and not like trying to solve some complicated language design issue, is it even worth spending so much time (18 days so far for this tiny PR) on it debating the space character? Is it ok to have a non 100% ideal change (from someone's point of view anyways) and then evaluate and fix if it's really not working out, in the launch-and-iterate spirit?
> 
> No one is suggesting to have a 100% ideal change, they're just suggesting to avoid adding another instance of complexity that doesn't make sense to them as a code reviewer. I think there is a pretty clear path forward here.
> 
> > I actually am now quite reluctant to start doing other/bigger PRs because I feel they might take forever because of some similar minor/arbitrary thing. @chandlerc for thoughts.
> 
> I'm sorry you're reluctant, but I do think there was already a pretty clear path forward here.
> 
> I think I've mentioned this before, but I feel like discussing planned changes ahead of time, or discussing disagreements in real time in Discord or one of the meetings might help speed things up. Moreover, I think if you want to make more incremental / fast progress, it may help to move longer discussions around changing the direction suggested in code review to not be in the critical path of the code review and instead be separate discussions.
> 
> Anyways, my two cents here.

Thanks Chandler. I guess my comment was more high level -- every single case usually has a 'way forward', but this is not the first time where a seemingly trivial PR I write results in a prolonged discussion which impacts velocity quite a bit, so I find it hard to justify based on an effort-vs-impact (and risk) basis. But maybe the approach here is different than what I am used to, or I'm misclassifying things as trivial whereas they are not? Not sure ;)

I went ahead and added a MakeKindName() method using the same .def machinery.